stdpopsim/ext/selection.py

```diff
@@ -151,3 +190,195 @@ def __attrs_post_init__(self):
+        raise ValueError("Selection coefficient must be non-negative.")
+
+    slim_events = [
+        DrawMutation(
+            single_site_id=single_site_id,
+            time=mutation_time,
+            population_id=population_id,
+        ),
+        ConditionOnAlleleFrequency(
+            single_site_id=single_site_id,
+            start_time=GenerationAfter(mutation_time),
+            end_time=end_time,
+            population_id=population_id,
+            allele_frequency=0.0,
+            op=">",
+        ),
+    ]
+
+    # Neutral period between mutation introduction and onset of sweep.
+    if mutation_time > start_time:
+        slim_events.extend(
+            [
+                ChangeMutationFitness(
+                    single_site_id=single_site_id,
+                    start_time=mutation_time,
+                    end_time=start_time,
+                    population_id=population_id,
+                    selection_coeff=0.0,
+                    dominance_coeff=dominance_coeff,
+                ),
+            ]
+        )
+
+    # Sweep period. Only check sweep AF conditions in start/end generations, so
+    # that allele trajectories are otherwise unconstrained.
+    slim_events.extend(
+        [
+            ChangeMutationFitness(
+                single_site_id=single_site_id,
+                start_time=start_time,
+                end_time=end_time,
+                population_id=population_id,
+                selection_coeff=selection_coeff,
+                dominance_coeff=dominance_coeff,
+            ),
+        ]
+    )
+
+    if min_freq_at_start is not None:
+        if not 0 < min_freq_at_start <= 1:
+            raise ValueError(
+                "If specified, the allele frequency at the start of the sweep "
+                "must be in (0, 1]."
+            )
+        if start_time == mutation_time:
+            raise ValueError(
+                "If the start time of the sweep coincides with the introduction "
+                "of the mutation, then `min_freq_at_start` cannot be set."
+            )
+        slim_events.extend(
+            [
+                ConditionOnAlleleFrequency(
+                    single_site_id=single_site_id,
+                    start_time=GenerationAfter(start_time),
+                    end_time=GenerationAfter(start_time),
+                    population_id=population_id,
+                    allele_frequency=min_freq_at_start,
+                    op=">=",
+                ),
+            ]
+        )
+
+    if min_freq_at_end is not None:
+        if not 0 < min_freq_at_end <= 1:
+            raise ValueError(
+                "If specified, the allele frequency at the end of the sweep "
+                "must be in (0, 1]."
+            )
+        slim_events.extend(
+            [
+                ConditionOnAlleleFrequency(
+                    single_site_id=single_site_id,
+                    start_time=end_time,
+                    end_time=end_time,
```

Comment:
above we had `GenerationAfter(start_time)`; should this also have `GenerationAfter( )`?

Reply:
I don't think so, because `end_time` can be 0. But you're right: when I look at when the mutation is increasing fitness, it's from `(start_time-1)` through `end_time`, and this doesn't align with the docstring.